pkg/cache/v3/simple_test.go

```diff
@@ -445,6 +460,177 @@ func TestSnapshotCreateWatchWithResourcePreviouslyNotRequested(t *testing.T) {
+	c := cache.NewSnapshotCache(false, cache.IDHash{}, logger{t: t})
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	if err := c.SetSnapshot(ctx, node.GetId(), snapshot); err != nil {
+		t.Fatalf("Failed to set snapshot: %v", err)
+	}
+
+	// Create a watchResponseCh for both listeners.
+	watchResponseCh := make(chan cache.Response, 1)
+	streamState := stream.NewStreamState(false, map[string]string{})
+	req := &discovery.DiscoveryRequest{
+		TypeUrl:       rsrc.ListenerType,
+		ResourceNames: []string{listenerName1, listenerName2},
+		Node:          node,
+	}
+	c.CreateWatch(req, streamState, watchResponseCh)
+
+	// Wait for the initial response with both listeners.
+	var gotWatchResponse cache.Response
+	var gotDiscoveryResponse *discovery.DiscoveryResponse
+	select {
+	case gotWatchResponse = <-watchResponseCh:
+		var err error
+		gotDiscoveryResponse, err = gotWatchResponse.GetDiscoveryResponse()
+		if err != nil {
+			t.Fatalf("Failed to get discovery response: %v", err)
+		}
+	case <-ctx.Done():
+		t.Fatal("Test timed out when waiting for a response from the watch")
+	}
+
+	// Validate the initial response.
+	gotVersion := gotDiscoveryResponse.GetVersionInfo()
+	if gotVersion != version {
+		t.Fatalf("Got version %q, want %q", gotVersion, version)
+	}
+	wantResources := map[string]types.ResourceWithTTL{
+		listenerName1: snapshot.Resources[types.Listener].Items[listenerName1],
+		listenerName2: snapshot.Resources[types.Listener].Items[listenerName2],
+	}
+	gotResources := cache.IndexResourcesByName(gotWatchResponse.(*cache.RawResponse).Resources)
+	if diff := cmp.Diff(wantResources, gotResources, protocmp.Transform()); diff != "" {
+		t.Fatalf("Response resources (-want +got):\n%s", diff)
+	}
+
+	// Simulate an ACK from Envoy by creating a watch for the same set of
+	// resources, but with the updated version and nonce from the initial
+	// response. The set of known resource names in the stream state is actually
+	// updated by the server implementation when it sees the ACK, before it
+	// recreates the watch.
+	streamState.SetKnownResourceNamesAsList(rsrc.ListenerType, []string{listenerName1, listenerName2})
+	req = &discovery.DiscoveryRequest{
+		TypeUrl:       rsrc.ListenerType,
+		ResourceNames: []string{listenerName1, listenerName2},
+		VersionInfo:   gotVersion,
+		ResponseNonce: gotDiscoveryResponse.GetNonce(),
+		Node:          node,
+	}
+	c.CreateWatch(req, streamState, watchResponseCh)
+
+	// Wait for the ACK to be processed. Rely on the stream state being updated
+	// with the known resource names.
+	wantKnownResourceNames := map[string]struct{}{
+		listenerName1: {},
+		listenerName2: {},
+	}
+	var gotKnownResourceNames map[string]struct{}
+	for ; ctx.Err() == nil; <-time.After(10 * time.Millisecond) {
+		gotKnownResourceNames = streamState.GetKnownResourceNames(rsrc.ListenerType)
+		if diff := cmp.Diff(wantKnownResourceNames, gotKnownResourceNames); diff == "" {
+			break
+		}
+	}
```

Comment:
Is this part important for the test?  I'm worried about the timing implications here indicating there is some kind of race.  Or is it the case that in practice these operations internally will be happening serially, and this is only needed because the test is trying to synchronize with the code under test?

Reply:
I was initially making the call to `c.CreateWatch` in a goroutine just like some of the existing tests. But then, I figured that that was not required, but forgot to update the code that subsequently checks for the stream state. Yes, there is no need to check for the stream state in a loop. Removed it to just check it inline.